EventGenerator/src/RPCGun_module.cc

```diff
@@ -56,214 +54,207 @@ namespace mu2e {
-          }else{
-            stop = tstop;
-            break;
+            else {
+              part = part->parent();
+              if ( std::binary_search(decayOffPDGCodes_.begin(), decayOffPDGCodes_.end(), int(part->pdgId()) ) ) {
+                tau += part->endProperTime() / gc.getParticleLifetime(part->pdgId()); 
+              }
           }
-        }
-      }
-    }else{
-      timemap->SetTime(protonPulse_->fire());
-      if (applySurvivalProbability_){
-        while (true){
-          const auto& tstop = stops_.fire();
-          double weight = exp(-tstop.tauNormalized)*survivalProbScaling_;
-          double rand = randomFlat_.fire();
-          if (weight > rand){
-            stop = tstop;
-            break;
-          }
-        }
-      }else{
-        const auto& tstop = stops_.fire();
-        stop = tstop;
+        } 
+        weight += exp(-tau);
       }
     }
+    if(doHistograms_) _time->Fill(weight);
+    return weight;
+  }
 
-    //std::cout << "Found stop " << exp(-stop.tauNormalized) << " " << stop.t << std::endl;
-
-    const CLHEP::Hep3Vector pos(stop.x, stop.y, stop.z);
-
-    const double energy = generateEnergy();
-
-    if(!generateInternalConversion_){
-      output->emplace_back( PDGCode::gamma,
-                            GenId::ExternalRPC,
-                            pos,
-                            CLHEP::HepLorentzVector( randomUnitSphere_.fire(energy), energy),
-                            stop.t );
-
-      event.put(std::move(output));
-      event.put(std::move(timemap));
-    } else {
+  //================================================================
+  void RPCGun::produce(art::Event& event) {
+    auto output{std::make_unique<StageParticleCollection>()};
+    const auto simh = event.getValidHandle<SimParticleCollection>(simsToken_); 
+    const auto pis = stoppedPiMinusList(simh);
+    double time_weight = MakeEventWeight(event);
+    std::unique_ptr<EventWeight> pw(new EventWeight(time_weight));
+    event.put(std::move(pw)); 
+    if(pis.empty()) {
+      throw   cet::exception("BADINPUT")
+        <<"RPCGun::produce(): no suitable stopped pion in the input SimParticleCollection\n";
+    }
+    for(const auto& pistop: pis) {
+      addParticles(output.get(), pistop);
+    }
+    event.put(std::move(output)); 
+   
+  }
+  
+  void RPCGun::addParticles(StageParticleCollection* output,
+                            art::Ptr<SimParticle> pistop)
+  {
+    //Photon energy and four mom:
+    double energy = spectrum_.sample(randSpectrum_.fire());
+    CLHEP::Hep3Vector p3 = randomUnitSphere_.fire(energy);
+    CLHEP::HepLorentzVector fourmom(p3, energy);
+    if(process_ == ProcessCode::ExternalRPC){
+     output->emplace_back(pistop,
+                         process_, 
+                         PDGCode::gamma,
+                         pistop->endPosition(),
+                         fourmom,
+                         pistop->endGlobalTime() + randExp_.fire(pionLifeTime_) 
```

Comment:
I don't think you need to add a decay time offset for RPC, due to the very short lifetime of stopped pions on a nucleus (I think at least)